src/com/sun/jna/win32/W32StringUtil.java

```diff
@@ -0,0 +1,390 @@
+     *
+     * @return The width of TCHAR in C
+     */
+    public static int getCharWidth() {
+        if (isAPITypeWide()) {
+            return Native.WCHAR_SIZE;
+        }
+        return 1;
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer) {
+        return toString(pointer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters) {
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @param lengthInCharacters The length of the string in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters, int lengthInCharacters) {
+        byte[] bytes = pointer.getByteArray(offsetInCharacters * getCharWidth(), lengthInCharacters * getCharWidth());
+        return toString(bytes, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @return The string
+     * @see #toString(byte[], long)
+     */
+    public static String toString(byte[] buffer) {
+        return toString(buffer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(byte[])
+     */
+    public static String toString(byte[] buffer, long offsetInCharacters) {
+        Memory pointer = new Memory(buffer.length + 1 * getCharWidth());
+        pointer.write(0, buffer, 0, buffer.length);
+
+        // Ensure null-termination
+        byte[] zeros = new byte[getCharWidth()];
+        Arrays.fill(zeros, (byte)0);
+        pointer.write(buffer.length, zeros, 0, zeros.length);
+
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @return The string
+     * @see #toString(char[], int)
+     * @see #toString(char[], int, int)
+     */
+    public static String toString(char[] buffer) {
```

Comment:
`Native#toString` a java char[] is always mapped to a wide char, and thus never native.

Reply:
Okay, now we are back to the other discussion again.
The simple reply to this statement is "NO". There are several functions in the code base of JNA where the buffer is a char[] BUT, the returned string can be either ANSI or UNICODE.  In the UNICODE case, you are right, but in the ANSI case, you are wrong. This method hides the fact that the char[] buffer should have been a byte[] buffer in order to handle both properly or even an instance of Memory.

An example:
```
public static String getTempPath() {  
  char[] buffer = new char[WinDef.MAX_PATH];  
  if (Kernel32.INSTANCE.GetTempPath(new DWORD(buffer.length), buffer).intValue() == 0) {  
    throw new Win32Exception(Kernel32.INSTANCE.GetLastError());  
  }  
  return W32StringUtil.toString(buffer);
}
```
The `buffer` variable will have the same type in ANSI or UNICODE, but in ANSI, only the first half of the variable will contain data.
In UNICODE, the string "Foo" would make the buffer contain: byte[] {70, 0, 111, 0, 111, 0, 0, 0}
In ANSI, the string "Foo" would make the buffer contain: byte[] {70, 111, 111, 0, 0, 0, 0, 0}

I do not see any other way to fix the problem other than doing this if an API breakage should be avoided.
I guess we could add a method that takes an instance of Memory instead and deprecate the method taking a char[], but this is breaking the API, right?

Reply:
So are you saying that `DWORD GetTempPath(DWORD nBufferLength, char[] buffer);` should have been `DWORD GetTempPath(DWORD nBufferLength, byte[] buffer);`? Forget about what we have implemented, what do you think the implementation for `DWORD GetTempPath` and `public static String getTempPath()` looked like if we were to restart?

Reply:
From a JNA consumer point of view, the best would be to offer a way to hide the different formats of a string (char* and wchar_t*).

```
private interface Kernel32_Test extends StdCallLibrary, WinNT, Wincon {
    Kernel32_Test INSTANCE = Native.load("kernel32", Kernel32_Test.class, W32APIOptions.DEFAULT_OPTIONS);
    DWORD GetTempPath(DWORD nBufferLength, Memory buffer);
}

public void getTempPath_alt1() {
    boolean isWide = W32APITypeMapper.DEFAULT == W32APITypeMapper.UNICODE;
    int charWidth = isWide ? Native.WCHAR_SIZE : 1;
    
    Memory buffer = new Memory(WinDef.MAX_PATH * charWidth);
    if (Kernel32_Test.INSTANCE.GetTempPath(new DWORD(buffer.size() / charWidth), buffer).intValue() == 0) {
        throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
    }
    
    if (isWide) {
        System.out.println(buffer.getWideString(0));
    } else {
        System.out.println(buffer.getString(0));
    }
}

public void getTempPath_alt2() {
    Memory buffer = W32StringUtil.allocateBuffer(WinDef.MAX_PATH);
    if (Kernel32_Test.INSTANCE.GetTempPath(new DWORD(W32StringUtil.toSizeInCharacters(buffer)), buffer).intValue() == 0) {
        throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
    }

    System.out.println(W32StringUtil.toString(buffer));
}
````

Looking at these 2 alternatives, I think alt1 puts too much code in the application that uses JNA. Sure, there are the utility classes, but then almost every function that retrieves a string need to have an utility method to do the glue.
Alt2 is cleaner for the consumer of the API, but I would really like to go one step further and sub-class Memory (or an alternative implementation that does not require native memory) that operates on characters rather than bytes. The class would have a method that could be used to fetch the number or characters that would fit in the allocated object, a simple getString() method that more or less does the same thing as W32StringUtil.toString(Pointer) is doing now.

I know that there is an API out there and we should try to avoid changing it unless absolutely needed, but I think this would make it easier for the users of JNA to do their application right.

In the win32 API, there are some functions that require a native char* in UNICODE and some that require wchar_t* in ANSI.
An example of the wchar_t* in ANSI is [NetUserAdd](https://docs.microsoft.com/en-us/windows/win32/api/lmaccess/nf-lmaccess-netuseradd).
An example of char* in UNICODE is [IIDFromString](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-iidfromstring)

The reason that I mentioned byte[] as a possible type for the parameter is that it's simply bytes. There is no way to know what kind of string it is until a charset is applied. Using char[] somewhat implies that the data is in UTF-16 format always, but that's not true in the ANSI case.



Another way to fix the issue with ANSI win32 API is to completely wipe support for ANSI and just provide the UNICODE version. As the support for ANSI is broken in so many different places in the JNA code, I guess it's unlikely anybody is using it.


So, I repeat what I've said before, there are 2 options:
1. Wipe ANSI support from JNA with minimal JNA API breakage for UNICODE win32 API
2. Fix ANSI support, but that will break JNA API even for UNICODE win32 API

Reply:
Please come to the mailing list - I raised that question there.

Reply:
If we wanted (2) I would support fixing `GetTempPath` and the likes to take `Memory`. This breaks loudly for anyone using JNA, so I wouldn't be worried. Subclassing `Memory` to memory that acts as a string seems ok too.